### ISSUE TITLE
[FEATURE] Provide a list of in-scope variables/locals as maps

### ISSUE THREAD

**Issue by skyzyx** [hashicorp/terraform#27715]
### Current Terraform Version

0.14.6

### Use-cases

The context is with using the New Relic Provider. Users provide the list of tag names and tag values to use, which get passed through to New Relic (via NRQL) which match a set of infrastructure to monitor. Users instantiate a list of tags and values in an `aws_resource_tags` module, which then gets passed into the New Relic module via a `tags` variable.

Inside the New Relic module, we pull out known-good tag/value pairings as local variables. Instead of having to re-type those same tag values a second time, I had wanted to enable people to refer to those same local variables which exist inside the module (DRY and all that) — but you can't do because the locals haven't been defined yet when users are instantiating the module. Even then, it creates an irresolvable circular reference. I get it.

So I started looking at implementing something like _late static binding_ using only HCL2 and Terraform. Users provide a marker/token which _represents_ the local variable on input/instantiation. The idea was that later, the module would replace these markers/tokens with real, resolvable values defined as locals. Simple for users, DRY, and mostly achievable. The idea was for them to be able to use a marker/token _any_ local variable, but I realized that there wasn't a good way to look them up without hard-coding a manual list.

I would love to be able to get a list of variables/locals from Terraform itself. I was thinking about something like `terraform.locals` and `terraform.variables` which reflect the existing HCL and provide a object (if we could support all variable types) or a map (if it were limited to a single type).

### Attempted Solutions

I chose a list of local variables to allow, and created a static mapping (instead of a dynamic mapping). Everything works as intended, except that the list of available variables is static and manual instead of dynamic and automatic.

### Proposal

I've dabbled with the `hcl/v2` package (e.g., `hclwrite`) and the `go-cty` package. There's _almost certainly_ a way to reflect over the HCL tree. Scalar values would definitely be simpler.

```hcl
# Normal usage (write)
locals {
  abc123 = {
    key = "value"
  }
}

# Normal usage (read)
local.abc123
local.abc123["key"]

# Equivalent for more dynamic use-cases (read-only)
terraform.locals["abc123"]["key"]
```

The variables/locals which exist when Terraform starts running are what would be available for runtime calculations. Functions like `try()` would work as expected.

### References

* https://www.php.net/manual/en/language.oop5.late-static-bindings.php

**Comment by apparentlymart** [comment]
Hi @skyzyx! Thanks for sharing this proposal.

I'm not sure I fully understood the use-case that prompted you to make this proposal, and so I don't feel able to evaluate this and other possible solutions to it. To help with discussion of this proposal, could you share the configuration you referred to in your "Attempted Solutions" section, where you suggested that it had worked but you wanted it to be "dynamic and automatic"?

**Comment by skyzyx** [comment]
Sure, and thank you for taking a look.

In short, there are some use cases where it would be helpful to be able to look up a variable's value by string. As in, my end-user gives me the name of a documented "local" variable they want to use the value of, and I look that value up and use it's value on their behalf.

While I think this would have genuine value to module developers with other use-cases besides mine, here is my use-case:

### Use-case

Instead of producing mega-modules that are monolithic and touch many things, my team and I have chosen to produce a series of smaller modules that work well together, with the idea of [function composition](https://en.wikipedia.org/wiki/Function_composition_(computer_science)) being a core tenet. [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) is another.

So we have one module for collecting the tags that represent a "Service" or "Workload".

```hcl
module "aws_resource_tags" {
  source      = "git::ssh://...../aws-resource-tags.git?ref=3.1.0"
  account     = "acctAAA"
  application = "devtools"
  environment = "prod"
  function    = "cluster"
  platform    = "core-services"
  runteam     = "torchwood"
}
```

The output of that module gets passed into all manner of other modules, depending on what people are building, and which pieces they need. It also works with the New Relic Alerts module we've built.

```hcl
module "nri_alerts" {
  source = "git::ssh://...../newrelic-alerts.git?ref=3.0.0"

  # Tags
  tags = module.aws_resource_tags.common_tags

  # …snip...
}
```

Now, the `nri_alerts` module knows which tags we should be paying attention to for this set of alerts. Every alert definition has a variety of parameters that can be passed-in as a map with optional keys (this was a tremendous amount of code to write before the optional-keys-in-maps experiment launched in Terraform 0.14!).

```hcl
module "nri_alerts" {
  source = "git::ssh://...../newrelic-alerts.git?ref=3.0.0"

  # Tags
  tags = module.aws_resource_tags.common_tags

  # An example alert to enable.
  cond_sqs_visible_messages = {
    create = true
    crit   = 5000
  }

  # Another example.
  cond_rds_max_conns_used_mysql = {
    create = true
    instance_type = "db.m5.xlarge"
    warn = 80 # %
  }
}
```

In the New Relic Alerts module, we use this inputs to construct the various values needed to pass into the [Terraform Provider for New Relic](https://registry.terraform.io/providers/newrelic/newrelic/). So far, so good.

### Customization

But teams have been requesting even more flexibility with how the underlying [NRQL](https://docs.newrelic.com/docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-new-relics-query-language) query is formatted and the things which are being queried. One of those areas is the tags (above) used in matching AWS resources to monitor — they want to mix and match to better fit their more nuanced needs. Perfectly understandable.

So, while we have sensible default values for constructing the NRQL query, I started thinking that I could use some combination of `format()`, `join()`, and maybe `regex()` to allow more flexibility. So I exposed a new `list(map)` variable which could accept something like this (example):

```hcl
module "nri_alerts" {
  source = "git::ssh://...../newrelic-alerts.git?ref=3.0.0"

  # …snip...

  resource_tags = [
    {
      tag   = "label.Account"
      value = "@@tag_account@@"
    },
    {
      tag   = "label.Application"
      value = "@@tag_application@@"
    },
    {
      tag   = "label.Platform"
      value = "MyExamplePlatform"
    },
  ]
}
```

The `@@` markers annotate these strings as something _replaceable_, e.g., like pseudo-variables. In this case, `@@tag_account@@` and `@@tag_application@@` say "use whatever the underlying tag value is". On the other hand, `MyExamplePlatform` is a literal string.

Inside the New Relic Alerts module, tags are extracted into locals for ease of working. In the case of `@@tag_application@@`, it's just a straight lookup from what we receive from the `aws_resource_tags` module. (A functional, but (IMO) ugly alternate would be `module.aws_resource_tags.application`.)

However, `@@tag_account@@` isn't a straight lookup in the case of the New Relic module (although it can be in some of the other modules). Here, we accept a "friendly" name and do some other calculations and lookups to get the value that New Relic needs (which is why I didn't build this into the upstream `aws_resource_tags` module). Therefore, using `module.aws_resource_tags.account` as input isn't _preferable_.

### Implementation

1. My first thought was to take this set of key-value pairs, use them to construct the NRQL query, then go back through and replace all of the `@@` markers with their variable values. For example, if someone wanted to use the internal `local.tag_account` variable (which is documented in the README as something that can be used), they would pass `@@tag_account@@` on the input. Ideally, we could provide access to _all_ internal variables, and people could just use what they want.

2. Then I set out to address taking the constructed NRQL query, and performing the replacements. I was going to use `regexall()` to grab the value inside the `@@` pairings, then lookup the value for that local variable (this ticket), and perform a series of consecutive replacements on the source string (#27716). No dice. :(

3. (Many hours of trial, error, searching the internet, repeat…)

4. I settled on a solution which was _good enough_, but not quite as flexible as I'd originally envisioned. Due to #27716 not existing yet, I refactored my solution to do the replacement first BEFORE `join()`ing the input into a single string. That solved that. I was hoping to implement a solution (for this ticket) _this way_ (dynamic access to variables):

    ```hcl
    locals {
      # …snip...

      resource_tags = join(" AND ", [
        for t in var.resource_tags : "${t.tag} = '${try(terraform.locals[t.value], t.value)}'"
      ])
    }
    ```

    …but I ended up having to solve it this way (static access to variables).

    ```hcl
    locals {
      # …snip...

      var_map = {
        "@@tag_account@@"          = local.tag_account
        "@@tag_application@@"      = local.tag_application
        "@@tag_business_service@@" = local.tag_business_service
        "@@tag_environment@@"      = local.tag_environment
        "@@tag_function@@"         = local.tag_function
        "@@tag_platform@@"         = local.tag_platform
        "@@tag_runteam@@"          = local.tag_runteam
      }

      resource_tags = join(" AND ", [
        for t in var.resource_tags : "${t.tag} = '${try(local.var_map[t.value], t.value)}'"
      ])
    }
    ```

### Why didn't I do it another way?

1. **Using `modules.*.*` syntax:** — For a _module_, it's possible that I could have reworked the solution to use more of this, and in _some_ cases, it would have been fairly clean. In other cases, however, I would have needed to expose/push implementation details onto the end-user that they shouldn't _need_ to concern themselves with. Enabling lookup by value name feels _cleaner_ for the end-user.

2. **Why not just use a static map? People shouldn't be able to access any internal like that!** — I did end up using a static map, and _maybe_ people shouldn't access anything they want. For my personal use-case, that _may_ be true.

    But as someone who writes complex software for other people to use, a majority of people will give-up if a thing doesn't work the way they think it should. [“For every customer complaint there are 26 other unhappy customers who have remained silent” – _Lee Resource_ (trying to find the source link to the source research)]. So I'm attempting to speak-up on behalf of them, anonymous as they are.

    If someone wants to use a variable that isn't "on the list", it requires pushing out a new release and/or them forking the module and falling out of sync with the upstream.

### Bottom line

I was able to work around my issue with a good-enough solution. This ticket is not intended to come up with other clever ways to solve my direct use-case (although that's generally much appreciated).

This ticket is specifically to request that the internal model of top-level variable names and values be exposed as an object, where the variable name can be looked-up by string, as I think there are use-cases for it in complex _modules_. You can certainly approve or decline this request, and I'm happy to discuss further, but the request is for the addition of this specific feature in a future version of Terraform.

Thank you!
